### ISSUE TITLE
how can we use varchar instead of nvarchar for models.CharField

### ISSUE THREAD

**Issue by hlacikd** [microsoft/mssql-django#175]
we have table that uses varchar as datatype, we are trying to use it as foreign key to another table completely created by models.Model

However mssql-django uses nvarchar as data type for models.CharField, which throws an error of non matching data types.

Question :

How to configure mssql-django with MSSQL server to use datatype varchar for models.CharField, or even how to create custom model which will lead to creating varchar datatype for collumn.

thanks in advance

**Comment by mShan0** [comment]
Hi @hlacikd, we're currently looking into this. Will get back to you once we have an optimal solution

**Comment by aashayamballi** [comment]
In our project also we're facing the same issue. Our parent table's primary has varchar datatype and when we're trying to migrate the Django models. we're getting the migration issue 

`django.db.utils.ProgrammingError: ('42000', "[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Column 'field' is not the same data type as referencing column 'field' in foreign key 'app_model_field_name_9a1ad41b_fk_table_field'. (1778) (SQLExecDirectW); [42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Could not create constraint or index. See previous errors. (1750)")`

In our case, the parent table was created directly in the SQL server and the model which I'm trying to create is through Django models and migrations.

**Comment by mShan0** [comment]
Hi @aashayamballi, can you post your models.py for your affected tables?

**Comment by aashayamballi** [comment]
> Hi @aashayamballi, can you post your models.py for your affected tables?

Sure i can share it tomorrow.. but, as I said the parent model was earlier created directly through DDL commands in SQL Server.. I just ran the inspectdb command to generate the django models from existing tables.

**Comment by aashayamballi** [comment]
@mShan0 this what django model looks like

```
class PreExistingModel(models.Model):
    field= models.CharField(
        db_column="db_field_name", primary_key=True, blank=True, null=False, max_length=10
    ) 
    field_1= models.TextField(db_column="db_field_name_1", blank=True, null=True) 
    
    class Meta:
        managed = False
        ordering = ["field_1"]
        db_table = "table_name"


class ModelCreatedByDjango(models.Model):
    class AccessStatus(models.TextChoices):
        APPROVE = "Approve", "Approved"
        IN_PROGRESS = "In progress", "In progress"
        REJECT = "Reject", "Rejected"

    user = models.ForeignKey(User, on_delete=models.CASCADE)
    m_2_m_field = models.ManyToManyField(PreExistingModel, blank=True)

```
so the first model PreExistingModel was created in SQL Server and we just ran inspectdb command and generated the DB model. so the PreExistingModel table's primary key type is varchar in the database.

so now when I'm trying to create the ManyToManyField in the second table, the through table is creating nvarchar field and when it's trying to add the refernce it is failing because the main table has varchar and the django's through table is created with nvarchar field.

**Comment by mShan0** [comment]
Thanks, continuing to look into this

**Comment by mathlesperance** [comment]
It would certainly be optimal to be able to specify if the underlying dbcolumn of a CharField IsUnicode (nvarchar) or not (varchar) for unmanaged table (tables created outside of Django).

Even in a simple case with no JOIN, not being able to specify that can have a big performance impact.

In our scenario we have primary keys and indexes on non-unicode varchar columns. 
When we look at the query generated, the parameters are always sent to MSSQL as nvarchar. 
This force MSSQL to do a Clustered Index Scan / Index Scan instead of a Clustered Index Seek / Index Seek.
This can change sub 10ms queries to 5/10/30+ seconds queries.

For now, our only solution is to use django ORM only for tables managed or created by Django and to use raw SQL for legacy tables created outside of Django.

Or am I missing something?

Thanks

**Comment by geminizb** [comment]
I think you need to config the database driver in django settings.py, like below
"OPTIONS": {
        "setencoding": [
            {"encoding": "utf-8"}
        ],
        ...
},

**Comment by tgs** [comment]
We're having the same Index Scan vs Index Seek problem as @mathlesperance, based on the parameter data type

**Comment by tgs** [comment]
The OPTIONS change did not change anything for me

**Comment by DrMeosch** [comment]
Thank you so much, @geminizb. In my case that helped me a lot!

**Comment by steveplanet** [comment]
YES! that fixed it by setting the encodings {"encoding": "utf-8"} in my DB settings (settings.py > DATABASES > 'mssql' > 'OPTIONS' ) FIXED a huge problem that I had where I was accessing VARCHAR strings in MSSQL and the execution plan was trying to use NVARCHAR so SQL was taking too long to do simple selects because on SQL, using my SQL profiler trace tool, I could see the SP were generated the passed parameters as `@P1 NVARCHAR(50)` and this fixed it by using `@P1 VARCHAR(50)` in my case, and fixed it!  thank you [geminizb](https://github.com/geminizb)
